### PR TITLE
Feature/add category order hint

### PR DESCRIPTION
**Pull request by PhilippSpo**

### CODE REVIEW

**Comment by PhilippSpo** [comment]
@sebbulon @hajoeichler @emmenko 
import:
- added mapping for the `categoryOrderHint` column, so the csv can get validated
- added validation for the `categoryOrderHint` column

export:
- added mapping for the `categoryOrderHint` property so it gets added to the csv

**Comment by mmoelli** [comment]
@PhilippSpo Please also add some info in the ReadMe to explain how to use the new column.

**Comment by PhilippSpo** [comment]
@mmoelli added docs for the new column. Are they clear enough?

**Comment by mmoelli** [comment]
looks great

**Comment by sebbulon** [comment]
@PhilippSpo question: Does it only work with the internal category id? Or can I also use the category name or category slug to identify my category in this field?

**Comment by PhilippSpo** [comment]
@sebbulon currently only internal category id and category name. But I already talked to @mmoelli to also use external category ids here.

File: src/coffee/exportmapping.coffee [hunk 1]
```diff
@@ -99,6 +99,13 @@ class ExportMapping
           else
             row[index] = product[attribName][lang]
 
+    if @header.has(CONS.HEADER_CATEGORY_ORDER_HINTS)
+      categoryIds = Object.keys product.categoryOrderHints
+      categoryOrderHints = []
+      _.each categoryIds, (categoryId) ->
+        categoryOrderHints.push("#{categoryId}:#{product.categoryOrderHints[categoryId]}")
+      row[@header.toIndex CONS.HEADER_CATEGORY_ORDER_HINTS] = categoryOrderHints.join ';'
```
Comment (hajoeichler) [inline src/coffee/exportmapping.coffee]: There are constants for the delimiters.

File: src/coffee/exportmapping.coffee [hunk 2]
```diff
@@ -99,6 +99,13 @@ class ExportMapping
           else
             row[index] = product[attribName][lang]
 
+    if @header.has(CONS.HEADER_CATEGORY_ORDER_HINTS)
+      categoryIds = Object.keys product.categoryOrderHints
+      categoryOrderHints = []
+      _.each categoryIds, (categoryId) ->
```
Comment (hajoeichler) [inline src/coffee/exportmapping.coffee]: Can't we use `_.map` here?

File: src/coffee/mapping.coffee [hunk 3]
```diff
@@ -112,6 +113,51 @@ class Mapping
...
+    # check if there actually is something to parse in the column
+    return catOrderHints unless @hasValidValueForHeader(rawMaster, CONS.HEADER_CATEGORY_ORDER_HINTS)
+    # parse the value to get a list of all catOrderHints
+    rawCatOrderHints = rawMaster[@header.toIndex CONS.HEADER_CATEGORY_ORDER_HINTS].split GLOBALS.DELIM_MULTI_VALUE
+    for rawCatOrderHint in rawCatOrderHints
+      # extract the category id and the order hint from the raw value
+      [rawCatId, rawOrderHint] = rawCatOrderHint.split ':'
+      orderHint = parseFloat(rawOrderHint)
+      # check if the product is actually assigned to the category
+      if _.has(@categories.id2fqName, rawCatId)
```
Comment (hajoeichler) [inline src/coffee/mapping.coffee]: I prefer the following style
```
catId = if ...
  rawcatId
else
 @categories...
```
```

File: src/coffee/mapping.coffee [hunk 4]
```diff
@@ -112,6 +113,51 @@ class Mapping
...
+        catId = rawCatId
```
Comment (hajoeichler) [inline src/coffee/mapping.coffee]: I guess this case is covered by the last `else` already

File: src/coffee/mapping.coffee [hunk 5]
```diff
@@ -112,6 +113,51 @@ class Mapping
...
     categories
 
+  # parses the categoryOrderHints column for a given row
+  mapCategoryOrderHints: (rawMaster, rowIndex) ->
+    catOrderHints = {}
+    # check if there actually is something to parse in the column
+    return catOrderHints unless @hasValidValueForHeader(rawMaster, CONS.HEADER_CATEGORY_ORDER_HINTS)
+    # parse the value to get a list of all catOrderHints
+    rawCatOrderHints = rawMaster[@header.toIndex CONS.HEADER_CATEGORY_ORDER_HINTS].split GLOBALS.DELIM_MULTI_VALUE
+    for rawCatOrderHint in rawCatOrderHints
```
Comment (hajoeichler) [inline src/coffee/mapping.coffee]: Please use the `underscore` library.

File: src/coffee/mapping.coffee [hunk 6]
```diff
@@ -112,6 +113,51 @@ class Mapping
...
+        else
+          @errors.push msg
+      else if !(orderHint > 0 && orderHint < 1)
+        msg = "[row #{rowIndex}:#{CONS.HEADER_CATEGORY_ORDER_HINTS}] Order hint has to be < 1 and > 0 but was '#{orderHint}'!"
+        if @continueOnProblems
+          console.warn msg
+        else
+          @errors.push msg
+
+      catOrderHints[catId] = rawOrderHint
```
Comment (hajoeichler) [inline src/coffee/mapping.coffee]: Even if the `orderHint` is `NAN`, you still store it here.

File: src/coffee/mapping.coffee (hunk 5 above)
Comment (PhilippSpo) [inline src/coffee/mapping.coffee]: I honestly just copy pasted this from the method above... Should I go for consistency or quality of new code?

**Comment by PhilippSpo** [comment]
@hajoeichler @emmenko @mmoelli @sebbulon I updated the version of the sdk and added more tests for this, also for using an external category id, category name and category slug. Are all requirements covered by the tests now?

**Comment by hajoeichler** [comment]
:+1: and thx for the test

**Comment by mmoelli** [comment]
looks good. 2 minor things:
* I haven't found a test for the export
* I'm missing the full explanation what kind of import options are supported

**Comment by PhilippSpo** [comment]
@mmoelli added test for exporting and added docs for import options

**Comment by mmoelli** [comment]
Looks great. :+1:

**Comment by sebbulon** [comment]
:+1:

**Pull request merged**